### PR TITLE
[stark] Extended SDF-stratified vol sampling (45 epochs) on corrected dataset

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

PR #972 SDF-stratified vol importance sampling is the current single-model SOTA on the corrected dataset (val_abupt=6.126%, test_abupt=5.844%). The 30-epoch schedule was inherited from pre-corrected-split tuning. Hypothesis: the corrected dataset has more genuine signal (50 test cases vs 34 val), and the loss curves for `56bcqp3m` were still declining when training stopped. Extending the SDF-stratified run to 45 epochs with a properly extended cosine schedule should push val_abupt below 6.0% — a key research milestone.

## Gate to Beat

- val_abupt < 6.126% (single-model SOTA, PR #972)
- val_vol_p < 3.798%

Stretch: val_abupt < 6.0% — first single-model run below the psychological 6.0% threshold on the corrected dataset.

## Instructions to Student

1. Reproduce PR #972's configuration but with **45 epochs** (vs 30) and an extended cosine schedule (T_max=45).
2. Key flags to use (copied/adapted from PR #972):
   - `--sdf-stratified-sampling` (or equivalent flag — check `python train.py --help`)
   - `--vol-points 65536`
   - All other model defaults from PR #972 single-model best (L=5, h=512, standard backbone)
   - Corrected dataset: `--data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511`
   - DDP8 across all 8 GPUs, batch_size per-GPU=4
3. **EP10 viability gate**: val_abupt ≤ 7.0% AND val_vol_p ≤ 4.5% (continue) ; > 8.0% or > 5.5% (kill).
4. **EP25 confirmation gate**: val_abupt ≤ 6.3% AND val_vol_p ≤ 4.0% (continue to 45) ; > 6.5% or > 4.5% (kill).
5. Save best-val checkpoint and report test metrics evaluated from best-val rather than terminal epoch.
6. Post terminal SENPAI-RESULT marker.

If you discover the PR #972 implementation has changed flag names or the exact SDF stratification arg has moved, ask the advisor in a PR comment with `python train.py --help` output before launching.

## Current Baseline Metrics (corrected dataset)

| Metric | Val | Test | Source |
|--------|-----|------|--------|
| abupt | 5.758% | 5.594% | PR #1059 K=4 ensemble |
| abupt (single-model) | 6.126% | 5.844% | PR #972 single-model SOTA |
| vol_p (single-model) | 3.798% | 3.643% | PR #972 |
| surf_p | — | 3.366% | PR #1059 ensemble |
| WSS (vec) | — | 6.330% | PR #1059 ensemble |

## Terminal Result Marker

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_axis_mean_rel_l2_pct","value":<number>},"test_metric":{"name":"test_abupt_axis_mean_rel_l2_pct","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Your pod has been assigned but no W&B run has started. Please launch the 45-epoch SDF-stratified run immediately per the PR body instructions. 

Use the corrected dataset and the SOTA config from PR #958 as your base. Key details:

- Base config: all flags from PR #958 tay-native SOTA (val_abupt=6.285%, val_vol_p=3.939%)
- Extended to 45 epochs with `--lr-cosine-t-max 45`
- Dataset: `--data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511`
- Add SDF-stratified sampling as in PR #972

Post the W&B run ID here as soon as training starts, then report EP3 metrics. The kill gate at EP3 is: val_abupt ≤ 8.0% AND val_vol_p ≤ 5.5%.

This is a high-priority experiment — PR #972's SDF-stratified approach is our current single-model SOTA and extending it to 45 epochs on the corrected dataset could compound those gains significantly.

**Comment by morganmcg1** [comment]
ADVISOR (correction to previous comment): Important clarification before you start.

The SDF-stratified sampling from PR #972 was implemented on the DL24 branch as a code change to `train.py` (monkey-patching + `--use-sdf-stratified-vol-sampling` and `--sdf-stratified-alpha` CLI args). The tay branch does **not** have these flags. You need to implement them on your branch.

Here are the exact implementation steps:

**Step 1: Add CLI args to `target/train.py` parser section:**
```python
parser.add_argument("--use-sdf-stratified-vol-sampling", action="store_true", default=False)
parser.add_argument("--sdf-stratified-alpha", type=float, default=2.0,
                    help="Bias strength: weight[i] = 1 + alpha * |sdf[i]|. Default 2.0.")
```

**Step 2: Add the monkey-patch block in `target/train.py`** after the training dataset is constructed but before the DataLoader is built. Copy the exact monkey-patch logic from PR #972's body (it's in the PR description under '### Implementation').

**Step 3: Launch the 45-epoch run:**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent stark --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 45 --epochs 45 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-sdf-stratified-vol-sampling --sdf-stratified-alpha 2.0 \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --wandb-group stark-sdf-stratified-45ep \
  --wandb-name stark/sdf-stratified-alpha2-45ep-v1
```

**Important flags NOT to use** (tay-incompatible, will crash):
- `--model-pe`, `--pe-init-sigmas`, `--use-y-symmetry-aug`, `--y-symmetry-aug-prob`, `--seed`, `--use-gradnorm` (use the flag names above instead)

**EP10 gate:** val_abupt ≤ 7.0% AND val_vol_p ≤ 4.5% (continue); > 8.0% OR > 5.5% (kill and report back).
**EP25 gate:** val_abupt ≤ 6.3% AND val_vol_p ≤ 4.0% (continue to 45); > 6.5% OR > 4.5% (kill).

Post W&B run ID here as soon as you launch. If you hit any issues with the monkey-patch implementation, post the error and I'll help debug.

**Comment by morganmcg1** [comment]
ADVISOR: Checking W&B — no runs found in group `stark-sdf-stratified-45ep`. Your pod has been idle.

My previous comment has the complete implementation instructions. Please start now:

1. Implement `--use-sdf-stratified-vol-sampling` and `--sdf-stratified-alpha` CLI args + the monkey-patch block in `target/train.py` (exact code is in the previous ADVISOR comment).

2. Launch the 45-epoch run:
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py   --dataset-path /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511   --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128   --batch-size 4   --optimizer lion --lr 9e-5 --weight-decay 5e-4   --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0   --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1   --pos-encoding-mode string_separable --use-qk-norm   --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0"   --lr-cosine-t-max 45 --epochs 45   --vol-points-schedule "0:16384:3:32768:6:49152:9:65536"   --no-compile-model   --train-surface-points 65536 --eval-surface-points 65536   --train-volume-points 65536 --eval-volume-points 65536   --use-surf-to-vol-xattn   --use-sdf-stratified-vol-sampling --sdf-stratified-alpha 2.0   --wandb-group stark-sdf-stratified-45ep   --wandb-name stark/sdf-stratified-alpha2-45ep-v1
```

**EP10 gate:** val_abupt ≤ 7.0% AND val_vol_p ≤ 4.5% (continue); > 8.0% OR > 5.5% (kill).
**EP25 gate:** val_abupt ≤ 6.3% AND val_vol_p ≤ 4.0% (continue); > 6.5% OR > 4.5% (kill).

Post the W&B rank-0 run ID immediately when launched. This is urgent — your pod has been idle.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION #4 — IMMEDIATE ACTION REQUIRED

You have received three prior escalations. Still zero W&B runs logged. You need to start this experiment NOW.

**Hypothesis**: SDF-stratified vol sampling biases training vol-point sampling toward near-surface regions, improving WSS accuracy.

**Implementation** (add to train.py vol-point sampling logic):

```python
# In the vol-point sampling section, replace uniform sampling with:
# 1. Compute SDF for each vol point (distance to nearest surface triangle)
# 2. Weight sampling probability as: w_i = exp(-sdf_i / sigma) where sigma=0.05
# 3. Sample vol points using these weights (torch.multinomial with replacement)
# sigma=0.05 means points within 5cm of surface are heavily upweighted
```

**Run command**:
```bash
python train.py   --wandb_group sdf-stratified-vol   --wandb_run_name stark-sdf-stratified-vol-v1   --pos-encoding-mode string_separable   --use-qk-norm   --rff-num-features 16   --rff-init-sigmas '0.25,0.5,1.0,2.0,4.0'   --vol-points-schedule '0:16384:3:32768:6:49152:9:65536'   --sdf-stratified-vol-sampling   --sdf-sigma 0.05   --lr-cosine-t-max 13   --batch_size 4
```

**EP1 kill gate**: val_abupt >30% at step 10,864 — stop and report. Otherwise continue.

Reply here with the W&B run ID as soon as it is running. If you are blocked by an implementation issue, describe it specifically and I will help unblock you.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION #5 — FINAL WARNING

This is the fifth escalation for PR #1065. You have received four prior escalations requesting you implement and launch the SDF-stratified vol sampling experiment. Still zero W&B runs in group `stark-sdf-stratified-45ep`.

**If no run is started within the next 2 hours, this PR will be closed as non-responsive and the GPU slot will be reassigned.**

Complete implementation and launch instructions were in my previous escalations. The key implementation is a monkey-patch to `data.py` adding SDF-biased sampling, then training with:

```bash
cd target/ && SENPAI_TIMEOUT_MINUTES=600 torchrun --standalone --nproc_per_node=8 train.py   --agent stark   --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511   --optimizer lion --lr 9e-5 --weight-decay 5e-4   --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0   --surface-loss-weight 2.0 --volume-loss-weight 1.0   --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1   --pos-encoding-mode string_separable --use-qk-norm   --rff-num-features 16 --rff-init-sigmas '0.25,0.5,1.0,2.0,4.0'   --use-sdf-stratified-vol-sampling --sdf-stratified-alpha 2.0   --lr-cosine-t-max 30 --epochs 30   --no-compile-model --model-layers 5 --model-hidden-dim 512   --model-heads 4 --model-slices 128 --batch-size 4   --train-surface-points 65536 --eval-surface-points 65536   --train-volume-points 65536 --eval-volume-points 65536   --use-surf-to-vol-xattn   --wandb-group stark-sdf-stratified-45ep   --wandb-name stark/sdf-stratified-alpha2.0
```

Post EP1 (step 10,864) and EP3 (step 32,592) metrics as PR comments.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1065 after 5 escalation warnings with zero W&B runs started. The 2-hour deadline issued at 07:53 UTC has passed. The GPU slot assigned to this experiment is being reassigned to a new hypothesis. If stark returns and wants to continue work on SDF stratified sampling, please open a new PR referencing this one.
